### DISCUSSION TITLE
string/equals

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#233] [Announcements]
# string/equals

在Java中，判断两个字符串是否相等需要注意不同的比较方法。本文详细介绍了如何使用equals()方法和==操作符来比较字符串，以及它们之间的区别和适用场景。掌握正确的字符串相等判断方法，避免常见的错误和陷阱。

https://tobebetterjavaer.com/string/equals.html

**Comment by Cantava** [comment]
"小萝莉" == "小" + "萝莉"拼接后
public String toString() {
    return new String(value, 0, count);
}
返回的结果是new出来的应该在堆里为什么会与“小萝莉‘’相等

**Reply by Tim1Wu** [reply]
文章这里是有误的，”小萝莉“是取得常量池的地址，”小“+”萝莉“只会创建一个堆对象，两者地址是不同的。

**Reply by 1274580448** [reply]
你们自己测试了吗。。。

**Reply by tough-road** [reply]
不都说了编译器优化了，把他俩拼接起来了，肯定一样，不用运行编译器就直接给你说这个一定是true

**Comment by ForeUP** [comment]
第五题的
```java
"小萝莉" == "小" + "萝莉"
// 输出 true
```
自动优化没解释清楚，因为[String、StringBuilder、StringBuffer](https://javabetter.cn/string/builder-buffer.html)这节里面并没有提到这种”自动优化“，根据我查到的内容，应该是属于**常量折叠**，导致输出为true，可以参考[知乎回答: Java中，关于String类型的变量和常量做“+”运算时发生了什么？](https://www.zhihu.com/question/35014775/answer/60785988)

**Reply by AHUA-Official** [reply]
+1     是编译器编译字节码文件的时候做的优化   知乎回答说的很对

**Reply by LanX1ang** [reply]
是的，我也很疑惑这个地方，去查了一下 这和StringBuilder没有关系，而是javac在编译期间的字面量折叠
